test/EngineTest.cpp

```diff
@@ -192,47 +330,111 @@ TEST(EngineTest, optionalJoinTest) {
-
-  ASSERT_EQ(3u, res.size());
-  ASSERT_EQ(inp[0], res[0]);
-  ASSERT_EQ(inp[2], res[1]);
-  ASSERT_EQ(inp[3], res[2]);
+  
+  // For easier checking.
+  IdTable sampleSolution = makeIdTableFromVector(std::vector<std::vector<size_t>>{{{1, 1, 3, 7}, {2, 2, 3, 5}, {3, 6, 5, 4}}});
+  ASSERT_EQ(sampleSolution.size(), res.size());
+  ASSERT_EQ(sampleSolution, res);
 }
+
+TEST(EngineTest, hashJoinTest) {
+  // Going through the tests of the normal lambda. The result for hashed join
+  // is sorted, if the input was also sorted, so hashJoin can go through them,
+  // without modifying them first.
+  Join J{Join::InvalidOnlyForTestingJoinTag{}};
+  auto HashJoinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    return J.hashJoin<A, B, C>(AD_FWD(args)...);
+  };
+  auto JoinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    return J.join<A, B, C>(AD_FWD(args)...);
+  };
+  goThroughSetOfTestsWithJoinFunction(createNormalJoinTestSet(), HashJoinLambda);
+
+  /*
+   * Create two sorted IdTables. Join them using the normal join and the
+   * hashed join. Compare.
+  */
+  std::vector<std::vector<size_t>> a = {
+      {34, 73, 92, 61, 18},
+      {11, 80, 20, 43, 75},
+      {96, 51, 40, 67, 23}
+    };
+ 
+  std::vector<std::vector<size_t>> b = {
+      {34, 73, 92, 61, 18},
+      {96, 2, 76, 87, 38},
+      {96, 16, 27, 22, 38},
+      {7, 51, 40, 67, 23}
+    }; 
+  
+  // Saving the results and comparing them.
+  IdTable result1 = useJoinFunctionOnIdTables(makeIdTableFromVector(a), 0, makeIdTableFromVector(b), 0, JoinLambda); // Normal.
+  IdTable result2 = useJoinFunctionOnIdTables(makeIdTableFromVector(a), 0, makeIdTableFromVector(b), 0, HashJoinLambda); // Hash.
+  ASSERT_EQ(result1, result2);
+
+  // Checking if result1 and result2 are correct.
+  compareIdTableWithSolution(result1, makeIdTableFromVector(std::vector<std::vector<size_t>>{
+      {
+        {34, 73, 92, 61, 18, 73, 92, 61, 18},
+        {96, 51, 40, 67, 23, 2, 76, 87, 38},
+        {96, 51, 40, 67, 23, 16, 27, 22, 38}
+      }
+    }),
+    true,
+    0
+  );
+  compareIdTableWithSolution(result2, makeIdTableFromVector(std::vector<std::vector<size_t>>{
+      {
+        {34, 73, 92, 61, 18, 73, 92, 61, 18},
+        {96, 51, 40, 67, 23, 2, 76, 87, 38},
+        {96, 51, 40, 67, 23, 16, 27, 22, 38}
+      }
+    }),
+    true,
+    0
+  );
+
+  // Random shuffle the tables, hash join them, sort the result and check,
+  // if the result is still correct.
+  randomShuffle(a.begin(), a.end());
+  randomShuffle(b.begin(), b.end());
+  result2 = useJoinFunctionOnIdTables(makeIdTableFromVector(a), 0, makeIdTableFromVector(b), 0, HashJoinLambda);
+
+  // Sorting is done inside this function.
+  compareIdTableWithSolution(result2, makeIdTableFromVector(std::vector<std::vector<size_t>>{
+      {
+        {34, 73, 92, 61, 18, 73, 92, 61, 18},
+        {96, 51, 40, 67, 23, 2, 76, 87, 38},
+        {96, 51, 40, 67, 23, 16, 27, 22, 38}
+      }
+    }),
+    false
+  );
+};
```

Comment:
This function still combines the definition of the test cases and running them.

There is one function missing:
````
runTestCaseForHashJoinAndZipperJoin(testCase) {
 // nothing in the testCase needs to be sorted
 // Random shuffle both tables, run hashJoin, check result.
 // Sort the larger table by join column, run hashJoin, check result (this time it's sorted).
// Sort both tables, run zipper join, check result.

Then you can put ALL your test cases for the joins (those defined here, and those from the "normalJoinTestCases" function into a single logic, and most of the code goes away.

Reply:
First off, calling the normal join function zipper join is bad form, because that is nowhere mentioned in the documentation and can cause rather huge misunderstandings. Double so, because zipper join is no the the standard name for that algorithm. (I still understand what you mean, but it is a dangerous habit.)

Secondly, I think creating a new function wouldn't be the best. There is currently no re usability in a function like that, so why not just make a new `TEST` instead? They do exist for separating different types of test cases.

Thirdly, I don't quite see, why zipper join should be a part of `runTestCaseForHashJoinAndZipperJoin(testCase)`. I mean, just put the new test cases together with the other test cases for zipper join after sorting them. What advantage would we gain, from doing it in `runTestCaseForHashJoinAndZipperJoin(testCase)`?
Even comparing the results of both join functions with each other falls flat, because we would already be comparing them both with the same sample solution. And by transitive property, they would be equal to each other, if they are equal to the sample solution.

Reply:
Alright, I see what you meant. Gotta implement that later, but slightly different.

Reply:
I wrote this function, but made it for a vector of `testCase`, because things like shuffling, or sorting, the input and if the result should be sorted, can be achieved by manipulating the `testCase`s. Which then allows me to 'run' the modified `testCase`s using the already written `goThroughSetOfTestsWithJoinFunction`, which saves a lot of work and time.

In light of this new function, I also got rid of `TEST(JoinTest, hashJoinTest)`, because it no longer served any purpose, after translating its additional tests into `testCase`s.

For the first draft of this function see [5e5a086](https://github.com/ad-freiburg/qlever/pull/821/commits/5e5a0862495e896dfc5a10d112db961205fcb4a9).